### PR TITLE
Update Snowflake doc

### LINKED ISSUE #2313: Document the AS CHANGELOG clause
This issue tracks the documentation update needed for the merged PR #17132.

Source PR URL: https://github.com/risingwavelabs/risingwave/pull/17132
Source PR Merged At: 2024-06-26T11:53:23Z

### LINKED ISSUE #2336: Document: fix(frontend): fix changelog out_col_change error and add snowflake upsert sink demo
This issue tracks the documentation update needed for the merged PR #17515.

Source PR URL: https://github.com/risingwavelabs/risingwave/pull/17515
Source PR Merged At: 2024-07-02T06:49:30Z

### LINKED ISSUE #2360: Document: refactor(sink): remove snowflake sink http client
This issue tracks the documentation update needed for the merged PR #17627.

Source PR URL: https://github.com/risingwavelabs/risingwave/pull/17627
Source PR Merged At: 2024-07-09T14:03:36Z

### PR DESCRIPTION
**Pull request by WanYixian**
## Info

- **Description**

  - Update Snowflake doc, see [preview](https://pr-2413.d2fbku9n2b6wde.amplifyapp.com/docs/next/sink-to-snowflake/).
  - Add `AS CHANGELOG` command, see [preview](https://pr-2413.d2fbku9n2b6wde.amplifyapp.com/docs/next/sql-as-changelog/) .

- **Notes**

  - [ Include any supplementary context or references here. ]

- **Related code PR**

  - [ Provide a link to the relevant code PR here, if applicable. ]

- **Related doc issue**
  
  Resolves https://github.com/risingwavelabs/risingwave-docs/issues/2336
  Resolves https://github.com/risingwavelabs/risingwave-docs/issues/2313
  Resolves https://github.com/risingwavelabs/risingwave-docs/issues/2360

## For reviewers

- **Preview**

  - [ Paste the preview link to the updated page(s) here. Edit this item after the preview site is ready. To find the updated pages, scroll down to locate and open the Amplify preview link and select the **dev** version of the documentation. ]

- **Key points**

  - [ Parts that may need revision or extra consideration. ]

## Before merging

- [ ] I have checked the doc site preview, and the updated parts look good.

- [ ] I have acquired the approval from the owner (and optionally the reviewers) of the code PR and at least one tech writer (`emile-00`, `hengm3467`, & `WanYixian`).

### CODE REVIEW

**Comment by WanYixian** [comment]
Hi @neverchanje, I've made some updates, could you pls continue working on it from here? Thanks a lot!

File: docs/guides/sink-to-snowflake.md [hunk 1]
```diff
@@ -11,21 +11,41 @@ slug: /sink-to-snowflake
...
-- For the table: The user must have either `OWNERSHIP` permission, or at the very least `INSERT` permission.
-- For the database: The user must have `USAGE` permission.
-- For the schema: The user must have `USAGE` permission.
+Meanwhile, the Snowflake user account must have the appropriate permissions:
+
+- For the pipe: The user must have `OPERATE`permission.
+- For the stage: The user must have `USAGE` and `READ` permission.
+- For the table: The user must have `INSERT` and `SELECT` permission.
+- For the database: The user must have `USAGE` permission.
+- For the schema: The user must have `USAGE` permission.
```
Comment (neverchanje) [inline docs/guides/sink-to-snowflake.md]: Regarding the permission setting, is there any snowflake doc that we can refer to users for details? @xxhZs

File: docs/guides/sink-to-snowflake.md [hunk 2]
```diff
@@ -146,3 +165,73 @@ CREATE SINK snowflake_sink FROM ss_mv WITH (
...
+    user_id,
+    target_id,
+    event_timestamp AT TIME ZONE 'America/Indiana/Indianapolis' as event_timestamp,
+    changelog_op AS __op,
+    _changelog_row_id::bigint AS __row_id
+FROM
+    sub;
+ ```
+ 
+Note that RisingWave uses `changelog` to transform streaming data into incremental logs. In the example above, `changelog_op` represents the type of modification (Insert/Update/Delete), while `_changelog_row_id` indicates the order of the modification. See [AS CHANGELOG](/sql/commands/sql-as-changelog.md).
```
Comment (neverchanje) [inline docs/guides/sink-to-snowflake.md]: BTW, I wonder if we can write some blogs on this feature.

**Comment by xxhZs** [comment]
snowflake sink has been refactored pr is
https://github.com/risingwavelabs/risingwave/pull/17627

File: docs/guides/sink-to-snowflake.md [hunk 3]
```diff
@@ -11,21 +11,41 @@ slug: /sink-to-snowflake
...
 
-Snowflake is a cloud-based data warehousing platform that allows for scalable and efficient data storage and analysis. For more information about Snowflake, see [Snowflake official website](https://www.snowflake.com/en/).
+Snowflake is a cloud-based data warehousing platform that allows for scalable and efficient data storage and analysis. For more information about Snowflake, see [Snowflake official website](https://www.snowflake.com/en/). 
+
+Sinking from RisingWave to Snowflake utilizes [Snowpipe](https://docs.snowflake.com/en/user-guide/data-load-snowpipe-rest-apis) for data loading. Initially, data is staged in a user-managed S3 bucket in JSON format, and then loaded into the Snowflake table via Snowpipe.
 
 ## Prerequisite
 
-- Ensure you already have a Snowflake table that you can sink data to. For additional guidance on creating a table and setting up Snowflake, refer to the [Getting started](https://docs.snowflake.com/en/user-guide-getting-started) guide of Snowflake documentation.
+- Ensure you have a S3 bucket that RisingWave can connect to.
```
Comment (emile-00) [inline docs/guides/sink-to-snowflake.md]:
```suggestion
- Ensure you have an S3 bucket that RisingWave can connect to.
```

File: docs/guides/sink-to-snowflake.md [hunk 4]
```diff
@@ -11,21 +11,41 @@ slug: /sink-to-snowflake
...
 
 - Ensure you have an upstream materialized view or source in RisingWave that you can sink data from.
 
+- Ensure you have a RSA key-value pair for authentication, and set up the credential for the `Snowflake User`.
```
Comment (emile-00) [inline docs/guides/sink-to-snowflake.md]:
```suggestion
- Ensure you have an RSA key-value pair for authentication, and set up the credential for the `Snowflake User`.
```

File: docs/guides/sink-to-snowflake.md [hunk 5]
```diff
@@ -11,21 +11,41 @@ slug: /sink-to-snowflake
...
+- Ensure you have a `Snowflake pipe` to receive the loaded data from the pre-defined stage and copy it to the Snowflake table.
+
+:::note
+RisingWave will not be responsible for deleting data already imported by S3. You can manually set the lifecycle configuration of your S3 bucket to clear out unnecessary data. See [Lifecycle configuration](https://docs.aws.amazon.com/AmazonS3/latest/userguide/how-to-set-lifecycle-configuration-intro.html) and [Delete staged files](https://docs.snowflake.com/en/user-guide/data-load-snowpipe-manage#deleting-staged-files-after-snowpipe-loads-the-datafor) for more details.
+:::
+
 ## Required permission
 
-To successfully sink data into Snowflake, the Snowflake user account must have the appropriate permissions. These permissions include:
+To successfully sink data into Snowflake, the S3 user account must have the appropriate permission:
```
Comment (emile-00) [inline docs/guides/sink-to-snowflake.md]:
```suggestion
To sink data into Snowflake, the S3 user account must have the appropriate permission:
```

File: docs/guides/sink-to-snowflake.md [hunk 6]
```diff
@@ -13,23 +13,25 @@ This guide describes how to sink data from RisingWave to Snowflake using the Sno
 
 Snowflake is a cloud-based data warehousing platform that allows for scalable and efficient data storage and analysis. For more information about Snowflake, see [Snowflake official website](https://www.snowflake.com/en/).
 
+Sinking from RisingWave to Snowflake utilizes [Snowpipe](https://docs.snowflake.com/en/user-guide/data-load-snowpipe-rest-apis) for data loading. Initially, data is staged in a user-managed S3 bucket in JSON format, and then loaded into the Snowflake table via Snowpipe. For more information, see [Overview of the Snowpipe REST endpoints to load data](https://docs.snowflake.com/user-guide/data-load-snowpipe-rest-overview).
```
Comment (xxhZs) [inline docs/guides/sink-to-snowflake.md:16]: Snowflake sink needs to write the data to s3 in json format first, then snowflake receives a notification from s3 and automatically ingests the data into the table. see [[Overview of the Snowpipe REST endpoints to load data](https://docs.snowflake.com/user-guide/data-load-snowpipe-rest-overview)] and [[Automating Snowpipe for Amazon S3](https://docs.snowflake.com/en/user-guide/data-load-snowpipe-auto-s3)].

File: docs/sql/commands/sql-as-changelog.md [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+id: sql-as-changelog
+title: AS CHANGELOG
+description: Convert stream into an append-only changelog. 
+slug: /sql-as-changelog
+---
+<head>
+  <link rel="canonical" href="https://docs.risingwave.com/docs/current/sql-as-changelog/" />
+</head>
+
+Use `AS CHANGELOG` clause to convert any stream into an append-only changelog. This feature is particularly useful when sinking data to append-only sinks like Snowflake using Snowpipe.
```
Comment (xxhZs) [inline docs/sql/commands/sql-as-changelog.md]: Converting a changlog op in stream into a column can be used to create mvs and sinks, a better practice is the upsert implementation of Snowflake sink

File: docs/sql/commands/sql-as-changelog.md [hunk 8]
```diff
@@ -0,0 +1,35 @@
...
+
+## Syntax
+
+```sql
+WITH table_name AS CHANGELOG FROM source_table;
```
Comment (xxhZs) [inline docs/sql/commands/sql-as-changelog.md:16]: This is done using the `cte` syntax, similar to 
WITH table_name AS CHANGELOG FROM source_table;
which converts the source_table change record to a column of table_name

File: docs/sql/commands/sql-as-changelog.md [hunk 9]
```diff
@@ -0,0 +1,35 @@
...
+```
+
+- The AS CHANGELOG clause is typically used when creating materialized views that need to be compatible with append-only sinks.
+- The resulting changelog includes additional columns such as `changelog_op` and `_changelog_row_id` to track changes.
+
+## Example
+
+```sql
```
Comment (xxhZs) [inline docs/sql/commands/sql-as-changelog.md]:
```create table user_behaviors(v1 int, v2 int);

  CREATE MATERIALIZED VIEW ss_mv AS
  WITH sub AS CHANGELOG FROM user_behaviors
  SELECT
      user_id,
      target_id,
      event_timestamp AT TIME ZONE 'America/Indiana/Indianapolis' as event_timestamp,
      changelog_op AS __op,
      _changelog_row_id::bigint AS __row_id
  FROM
      sub;
      
insert into user_behaviors values(1,1);

insert into user_behaviors values(2,2);

delete from user_behaviors  where v1 = 2;\

update user_behaviors  set v2 = 100 where v1 = 1;

select * from ss_mv
----------------------
v1 v2 changelog_op
1   1    1
2    2    1
2    2    2
1    1    4
1   100  3
```

File: docs/guides/sink-to-snowflake.md (hunk 6 above)
Comment (WanYixian) [reply docs/guides/sink-to-snowflake.md:16]: Yes, already mentioning [Automating Snowpipe for Amazon S3] in line 30 and line 101

**Review by xxhZs — APPROVED** [review]

**Pull request merged**